### PR TITLE
Add "iface-id-ver=${POD_UID}" tuple to the external-ids of logical and OVS ports

### PR DESCRIPTION
**Pull request by npinaeva**
Newly added option [1] allows to distinguish new vs old logical
ports for recreated pod with the same name, thus making "ovn-installed"
option reliable again [2].

1. Add "iface-id-ver" option to logical and OVS ports
[Upgrade][master] add iface-id-ver only for newly created Ports,
no ovn version check, since if iface-id-ver is not enabled in ovn, it will be just ignored
[Upgrade][node] no additional actions are needed, because existing Interfaces don't need
iface-id-set since it won't be set in master. Ovn support for iface-id-ver will be checked
during cni server creation
2. Add OvnKubeNodeConfig option "ovn-iface-id-ver-disabled" to manually
   disable this feature for older ovn versions
3. Change test mock that checks ovs add-port command
4. Add common function to compose LogicalPort name = iface-id
5. Enable OVNPortUp support after upgrade based on OVNTopo version and getOVNIfUpCheckMode check
6. revert [2]
[1] ovn-org/ovn@922c45f
[2] 22bed6a

Signed-off-by: Nadia Pinaeva <npinaeva@redhat.com>

### CODE REVIEW

**Comment by dcbw** [comment]
Could you do the revert as a separate commit, to keep the new changes smaller and more readable?

File: go-controller/pkg/cni/ovs.go [hunk 1]
```diff
@@ -307,12 +295,14 @@ func waitForPodInterface(ctx context.Context, mac string, ifAddrs []*net.IPNet,
 			if err := isIfaceIDSet(ifaceName, ifaceID); err != nil {
```
[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/pkg/cni/ovs.go:295]: @flavio-fernandes good idea, it could be a quicker check if a second sandbox has started after us. If we have already set iface-id-ver (which by this point we have) and then we re-read it and it has changed, that means kubelet/CRIO triggered a new sandbox request, and this one should exit.

**Comment by dcbw** [comment]
The only major concern I have is how we detect whether OVN supports iface-id-ver. Downstream in OpenShift we don't have that problem of course, since we have known OVN versions. But here upstream, we try to detect features and enable them optionally if they exist. That's what we do with ovn-installed itself (and we go through some icky heroics to do that). Can you doube-check with Numan if ther'es a good way to know that OVN supports this or not?

**Comment by npinaeva** [comment]
> The only major concern I have is how we detect whether OVN supports iface-id-ver. Downstream in OpenShift we don't have that problem of course, since we have known OVN versions. But here upstream, we try to detect features and enable them optionally if they exist. That's what we do with ovn-installed itself (and we go through some icky heroics to do that). Can you doube-check with Numan if ther'es a good way to know that OVN supports this or not?

Nice thought, I will check that, but I wonder why don't we know ovn version in upstream? Is it just not implemented or there are some reasons?
Enabling `ovn-installed` really looks very complicated.

File: go-controller/pkg/cni/ovs.go [hunk 2]
```diff
@@ -172,14 +172,28 @@ func getIfaceOFPort(ifaceName string) (int, error) {
...
+		tables []int
+	}
 
-func getLegacyFlowQueries(mac string, ifAddrs []*net.IPNet, ofPort int) []openflowQuery {
 	// Query the flows by mac address for in_port_security and OF port
-	queries := getMinimalFlowQueries(mac, ofPort)
+	queries := []query{
+		{
+			match:  "dl_src=" + mac,
+			tables: []int{9},
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/cni/ovs.go:190]: @flavio-fernandes : Table 9 is the port-security table. That is the last action that is done by the master in the `addLogicalPort` codeflow. So if we know that port security flows are present, we can be sure the ovnkube-master logic was successful.

[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/cni/ovs.go:190]: Can this `table number: meaning` info be found somewhere?

**Comment by numansiddique** [comment]
> > The only major concern I have is how we detect whether OVN supports iface-id-ver. Downstream in OpenShift we don't have that problem of course, since we have known OVN versions. But here upstream, we try to detect features and enable them optionally if they exist. That's what we do with ovn-installed itself (and we go through some icky heroics to do that). Can you doube-check with Numan if ther'es a good way to know that OVN supports this or not?
> 
> Nice thought, I will check that, but I wonder why don't we know ovn version in upstream? Is it just not implemented or there are some reasons?
> Enabling `ovn-installed` really looks very complicated.

Unfortunately there is no mechanism now.

File: go-controller/pkg/cni/ovs.go [hunk 3]
```diff
@@ -172,14 +172,28 @@ func getIfaceOFPort(ifaceName string) (int, error) {
 	return iPort, nil
 }
 
-type openflowQuery struct {
-	match  string
-	tables []int
-}
+func doPodFlowsExist(mac string, ifAddrs []*net.IPNet, ofPort int) bool {
+	// Function checks for OpenFlow flows to know the pod is ready
+	// TODO(trozet): in the future use a more stable mechanism provided by OVN:
```
Comment (astoycos) [inline go-controller/pkg/cni/ovs.go:177]: I think we can remove this TODO now Right @trozet ?

File: go-controller/pkg/cni/ovs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply go-controller/pkg/cni/ovs.go:190]: I would add a new const in the `config` package that has a comment + naming which reflects the importance of table 9

[the remark this answers is not in the document]
Comment (astoycos) [reply go-controller/pkg/cni/ovs.go:190]: `ovsPortSecTable` or something

File: go-controller/pkg/cni/ovs.go [hunk 4]
```diff
@@ -172,14 +172,28 @@ func getIfaceOFPort(ifaceName string) (int, error) {
 	return iPort, nil
 }
 
-type openflowQuery struct {
-	match  string
-	tables []int
-}
+func doPodFlowsExist(mac string, ifAddrs []*net.IPNet, ofPort int) bool {
```
Comment (astoycos) [inline go-controller/pkg/cni/ovs.go:175]: Wait why do we still need to manually make sure the ovs physical rules are there if `ovn-installed` works again?

**Comment by npinaeva** [comment]
So I added a flag to disable `iface-id-ver` usage manually, but I didn't provide any means to pass this option.
Also not sure if I should document it somewhere.
So please share your thoughts about naming/description/what I missed for this flag.
Revert for https://github.com/ovn-org/ovn-kubernetes/commit/22bed6a10c669142fb13e612a8fe17cf8b895fd6 will be done in the next PR

File: go-controller/pkg/config/config.go [hunk 5]
```diff
@@ -351,8 +351,9 @@ type HybridOverlayConfig struct {
 
 // OvnKubeNodeConfig holds ovnkube-node configurations
 type OvnKubeNodeConfig struct {
-	Mode           string `gcfg:"mode"`
-	MgmtPortNetdev string `gcfg:"mgmt-port-netdev"`
+	Mode                  string `gcfg:"mode"`
+	MgmtPortNetdev        string `gcfg:"mgmt-port-netdev"`
+	OVNIfaceIdVerDisabled string `gcfg:"ovn-iface-id-ver-disabled"`
```
Comment (dcbw) [inline go-controller/pkg/config/config.go:356]: @npinaeva I would use `disable-ovn-iface-id-ver`

File: go-controller/pkg/cni/ovs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/cni/ovs.go]: In CNI sandbox is associates with CNI_NETNS, and in the code PodRequest.SandboxID is set to CNI_CONTAINERID.
When you say "kubelet/CRIO triggered a new sandbox request" does it mean CNI request with the same K8S_POD_* arguments, but different CNI_CONTAINERID?

[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/cni/ovs.go]: If so, `iface-id-ver` will never change for given `ifaceName`, but `iface-id` will be cleaned by the next ADD request

[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/pkg/cni/ovs.go]: @npinaeva yes, that means same pod name + namespace but different NETNS. Our `iface-id-ver` should be the pod UUID, which *is* unique between requests.

**Comment by dcbw** [comment]
> So I added a flag to disable `iface-id-ver` usage manually, but I didn't provide any means to pass this option.

For OpenShift we don't need that means; flags get passed in the CNO but since we know the OVN version we're using supports iface-id-ver, we don't need to pass anything.

**Comment by dcbw** [comment]
This LGTM for now, we can clean up the cruft around checking for ovn-installed/iface-id-ver later and hide that behind the option.

File: go-controller/pkg/cni/ovs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/cni/ovs.go]: Yes, but what about CNI_CONTAINERID? because `ifaceName` is based on this variable

File: go-controller/pkg/config/config.go [hunk 6]
```diff
@@ -1072,6 +1073,12 @@ var OvnKubeNodeFlags = []cli.Flag{
 		Value:       OvnKubeNode.MgmtPortNetdev,
 		Destination: &cliConfig.OvnKubeNode.MgmtPortNetdev,
 	},
+	&cli.StringFlag{
+		Name:        "disable-ovn-iface-id-ver",
+		Usage:       "if iface-id-ver option is not enabled in ovn, set this flag to True (depends on ovn version)",
```
Comment (squeed) [inline go-controller/pkg/config/config.go:1078]: suggestion: list minimum OVN version in the Usage.

**Comment by squeed** [comment]
minor doc suggestion, otherwise lgtm.

File: go-controller/pkg/config/config.go (hunk 6 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go:1078]: In this case, it would be a git commit or 21.09 :)

Comment (npinaeva) [reply go-controller/pkg/config/config.go:1078]: Well I can comment 21.09 version as minimal? I guess it's better than nothing

Comment (squeed) [reply go-controller/pkg/config/config.go:1078]: yah some kind of useful version would be good.

**Comment by npinaeva** [comment]
@squeed  can you check the changes and run this workflow if everything is ok please?

File: go-controller/pkg/cni/ovs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/cni/ovs.go]: @dcbw it's important for my next revert PR ^

**Comment by squeed** [comment]
Looks good. I don't have the bits to kick off CI either.

**Comment by dcbw** [comment]
@npinaeva I think we need one more thing; which is that the master needs to set some flag (like `enable-iface-id-ver`) in the pod annotations which the node side reads and then selectively turns iface-id-ver on or off. The problem is upgrades; we upgrade nodes before masters so there is a window where the node might be looking for iface-id-ver to be present, but the old running master won't set it.

**Comment by trozet** [comment]
@npinaeva you can use this mechanism to detect that masters have been upgraded and then configure the CNI to support the new capability:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go#L479

Basically we have topoversions that we can increment. During upgrade, the node will wait for master to upgrade to that version, then enable iface-id-ver.

File: go-controller/pkg/ovn/pods.go [hunk 7]
```diff
@@ -285,6 +285,13 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
 		opts = make(map[string]string)
 	}
 	opts["requested-chassis"] = pod.Spec.NodeName
+	// Unique identifier to distinguish interfaces for recreated pods, also set by ovnkube-node
+	// ovn-controller will claim the OVS interface only if external_ids:iface-id
+	// matches with the Port_Binding.logical_port and external_ids:iface-id-ver matches
+	// with the Port_Binding.options:iface-id-ver. This is not mandatory.
+	// If Port_binding.options:iface-id-ver is not set, then OVS
+	// Interface.external_ids:iface-id-ver if set is ignored.
+	opts["iface-id-ver"] = string(pod.UID)
```
Comment (trozet) [inline go-controller/pkg/ovn/pods.go:294]: per our slack discussion, just move this up to the if lsp==nil block on 266. Then only new pods will have this set. We dont need to worry about upgrade because ovn-controller is backwards compatible if it doesnt see the new option on the LSP it will bind without checking for it in ovsdb column.

**Comment by npinaeva** [comment]
When I found the 3rd different way in code to compose iface-id/portname by concatenating strings, I created a new function for this purpose.
If you don't think it should be included in this PR, let me know

File: go-controller/pkg/node/node.go [hunk 8]
```diff
@@ -469,10 +475,6 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 						klog.Fatalf("Failed to upgrade service route for node, error: %v", err)
 					}
 				}
-				// ensure CNI support for port binding built into OVN, as masters have been upgraded
```
Comment (trozet) [inline go-controller/pkg/node/node.go:472]: why remove this?

File: go-controller/pkg/config/config.go [hunk 9]
```diff
@@ -1072,6 +1073,13 @@ var OvnKubeNodeFlags = []cli.Flag{
 		Value:       OvnKubeNode.MgmtPortNetdev,
 		Destination: &cliConfig.OvnKubeNode.MgmtPortNetdev,
 	},
+	&cli.StringFlag{
```
Comment (trozet) [inline go-controller/pkg/config/config.go:1076]: should be a bool

File: go-controller/pkg/util/util.go [hunk 10]
```diff
@@ -261,3 +261,21 @@ ipLoop:
 
 	return out
 }
+
+func GetLogicalPortName(podNamespace, podName string) string {
```
Comment (trozet) [inline go-controller/pkg/util/util.go:265]: ideally the cosmetic changes to creating the port name should be a separate commit in this PR. Will make backporting easier.

**Review by trozet — CHANGES_REQUESTED** [review]
some comments inline. I think you still need another commit to revert the "adds back checking for openflows" commit: https://github.com/ovn-org/ovn-kubernetes/pull/2220/commits/22bed6a10c669142fb13e612a8fe17cf8b895fd6

File: go-controller/pkg/node/node.go (hunk 8 above)
Comment (npinaeva) [reply go-controller/pkg/node/node.go:472]: This function enables ovn-installed usage based on ovn-kube version, but it should consider ovn version, which is checked here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/node.go#L251, and this flag is passed to NewCNIServer, and from that point it shouldn't change.

Comment (npinaeva) [reply go-controller/pkg/node/node.go:472]: @trozet  does that make sense?

Comment (trozet) [reply go-controller/pkg/node/node.go:472]: So the check you referenced on the CNI side does only check during ovnkube node start time, and its checking an entry in sbdb: 
```
	if _, stderr, err := util.RunOVNSbctl("--columns=up", "list", "Port_Binding"); err != nil {
		if strings.Contains(stderr, "does not contain a column") {
			klog.Infof("Falling back to using legacy CNI OVS flow readiness checks")
			return false, nil
		}
```

However, during upgrade ovnkube-node and ovn-controller will upgrade first. That means ovnkube-node will start and check SBDB which is not upgraded and does not have this column and think it does not support ovn-installed. Then when master upgrades, it will update the topo version and the node will read it here and enable the support.

I think this code needs to stay. However you point out that we assume if you are running this version of ovn-kubernetes then you have an OVN that is new enough to support ovn-installed. To fix that you can simply add ```getOVNIfUpCheckMode() ``` here as well. So:

```
if initialTopoVersion < types.OvnPortBindingTopoVersion && cniServer != nil  && getOVNIfUpCheckMode() {
```

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/node/node.go (hunk 8 above)
Comment (npinaeva) [reply go-controller/pkg/node/node.go:472]: Ok now I get it, upgrade order is not obvious, thanks for explanation!

**Review by trozet — APPROVED** [review]
/lgtm
